### ISSUE TITLE
bug: @DgsEntityFetcher does not delegate error handling correctly when using Mono

### ISSUE THREAD

**Issue by StefanVanDyck** [Netflix/dgs-framework#966]
## Expected behavior
When implementing an entity fetcher using `@DgsEntityFetcher` I would expect all errors to be handled by our custom exception handler, implementing the `DataFetcherExceptionHandler` interface.

## Actual behavior
Errors are not delegated to the exception handler.
Unless the errors are first mapped to `InvocationTargetException`.

## Steps to reproduce
Implementing two entity-fetchers like so, should demonstrate the behavior.
This also happens when returning a CompletableFuture and using `toFuture()` on the stream before returning.
```
  @DgsEntityFetcher(name = "User")
  public Mono<User> entitiesUserNotHandled(Map<?, ?> values, DgsDataFetchingEnvironment dfe) {
    return Mono.error(new RuntimeException("TEST"))
  }
  
  @DgsEntityFetcher(name = "User")
  public Mono<User> entitiesUser(Map<?, ?> values, DgsDataFetchingEnvironment dfe) {
    return entitiesUserHandled(Map<?, ?> values, DgsDataFetchingEnvironment dfe)
        .onErrorMap(InvocationTargetException::new);
  }
```

**Comment by StefanVanDyck** [comment]
Found the `onErrorMap` solution by stepping through the code and encountering this line:
https://github.com/Netflix/dgs-framework/blob/ac8c708352b3c58332c91bbe4feb18d3e45df078/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/federation/DefaultDgsFederationResolver.kt#L127

But I'm not entirely sure about what I am looking at :)

**Comment by srinivasankavitha** [comment]
Which version are you using? Have you tried using the latest release?

**Comment by StefanVanDyck** [comment]
I upgraded to the latest(at this time) version: 4.9.25

**Comment by denniseffing** [comment]
We encountered the same issue. Are there any known workarounds?

**Comment by StefanVanDyck** [comment]
Adding the `.onErrorMap(InvocationTargetException::new)` like in the example above should work.

**Comment by srinivasankavitha** [comment]
Is this an issue with the default exception handler or only when you are using a custom one? Because it is required to handle InvocationTargetException in the case of entities. I will close the issue in that case.

**Comment by StefanVanDyck** [comment]
@srinivasankavitha The issue is that when using a custom exception handler, 
exceptions within an EntityFetchers (unlike methods annotated with data, mutation or query) are not handled by that handler.
Unless they are explicitly wrapped inside of an InvocationTargetException.

**Comment by denniseffing** [comment]
To elaborate further:
The issue is not only that _custom_ exception handler are not called, the issue is that generally, the exception handler is not called at all when Monos are used.

This issue is caused by `fetcher.second.invoke` not throwing an `InvocationTargetException` that wraps the initially thrown exception. It returns a `Mono.error` that wraps the initial exception instead. This is caused by Mono never actually _throwing_ an exception that is caught by the Java API. As far as the Java API is concerned, the reflection `invoke` call completed normally (it just returned `Mono.error`). This in turn leads to normal code execution of the rest of the method.
https://github.com/Netflix/dgs-framework/blob/2210a86d6659d260921372536625ee9f8333cf3d/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/federation/DefaultDgsFederationResolver.kt#L91-L95

In this case, the if condition that checks for an `InvocationTargetException` fails and the else block is executed instead.
https://github.com/Netflix/dgs-framework/blob/2210a86d6659d260921372536625ee9f8333cf3d/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/federation/DefaultDgsFederationResolver.kt#L127

The else block however _does not delegate_ to the exception handler, but returns a statically generated `TypedGraphQLError` instead.
https://github.com/Netflix/dgs-framework/blob/2210a86d6659d260921372536625ee9f8333cf3d/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/federation/DefaultDgsFederationResolver.kt#L148

**Comment by raptros** [comment]
I think that denniseffing's explanation also explains why it affects CompletableFuture errors in the same way.

**Comment by srinivasankavitha** [comment]
Yes if its in a custom exception handler, you do need to check for
InvocationTargetException in the case of handling for entities.
